File: src/main/java/net/slipp/web/tag/AdminTagController.java
package net.slipp.web.tag;

import javax.annotation.Resource;

import net.slipp.domain.tag.Tag;
import net.slipp.service.tag.TagService;

import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.domain.Sort.Direction;
import org.springframework.stereotype.Controller;
import org.springframework.ui.ModelMap;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;

@Controller
@RequestMapping("/admin")
public class AdminTagController {
	private static final int DEFAULT_PAGE_NO = 1;
	private static final int DEFAULT_PAGE_SIZE = 20;

	@Resource(name = "tagService")
	private TagService tagService;
	
	@RequestMapping(value = "/tags", method = RequestMethod.GET)
	public String tags(Integer page, ModelMap model) throws Exception {
		model.addAttribute("tags", tagService.findAllTags(createPageable(page)));
		model.addAttribute("parentTags", tagService.findPooledParentTags());
		return "admin/tags";
	}

	Pageable createPageable(Integer page) {
		if (page == null) {
			page = DEFAULT_PAGE_NO;
		}
		Sort sort = new Sort(Direction.DESC, "tagId");
		Pageable pageable = new PageRequest(page - 1, DEFAULT_PAGE_SIZE, sort);
		return pageable;
	}
	
	@RequestMapping(value = "/tags", method = RequestMethod.POST)
	public String create(String name, Long parentTag, ModelMap model) throws Exception {
		Tag originalTag = tagService.findTagByName(name);
		if (originalTag != null) {
			model.addAttribute("errorMessage", name + " 태그는 이미 존재합니다.");
			return tags(1, model);
		}
		
		if (parentTag == null) {
			tagService.saveTag(Tag.pooledTag(name));
			return "redirect:/admin/tags";
		}
		
		Tag parent = tagService.findTagById(parentTag);
		if (parent == null) {
			model.addAttribute("errorMessage", parentTag + " 태그 아이디는 존재하지 않습니다.");
			return tags(1, model);
		}
				
		tagService.saveTag(Tag.pooledTag(name, parent));
		return "redirect:/admin/tags";
	}
	
	@RequestMapping(value = "/moveNewTag", method = RequestMethod.POST)
	public String moveNewTag(Long tagId, Long parentTag) throws Exception {
		tagService.moveToTag(tagId, parentTag);
		return "redirect:/admin/tags";
	}
}

File: src/main/scala/net/slipp/web/tag/AdminTagController.scala
package net.slipp.web.tag

import java.lang.Long

import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import com.typesafe.scalalogging.LazyLogging
import javax.annotation.Resource
import net.slipp.service.tag.TagService
import net.slipp.web.QnAPageableHelper.createPageableTagId
import org.springframework.web.bind.annotation.RequestMethod

@Controller
@RequestMapping(Array("/admin"))
class AdminTagController(
  @Resource(name = "tagService") tagService: TagService) extends LazyLogging {
  private val DefaultPageNo = 1
  private val DefaultPageSize = 20

  @RequestMapping(Array("/tags"))
  def tags(page: Integer, model: Model) = {
    tagsToModel(page, model)
    "admin/tags"
  }

  private def tagsToModel(page: Integer, model: Model) {
    model.addAttribute("tags", tagService.findAllTags(createPageableTagId(page, DefaultPageSize)))
    model.addAttribute("parentTags", tagService.findPooledParentTags)
  }

  @RequestMapping(value = Array("/tags"), method = Array(RequestMethod.POST))
  def create(@RequestParam name: String, @RequestParam parentTag: java.lang.Long, model: Model): String = {
    logger.debug(s"create tag : $name, $parentTag")

    try {
      tagService.saveTag(name, Option(parentTag))
      "redirect:/admin/tags"
    } catch {
      case e: IllegalArgumentException => {
        model.addAttribute("errorMessage", e.getMessage)
        tagsToModel(DefaultPageNo, model)
        return "admin/tags"
      }
    }
  }

  @RequestMapping(value = Array("/moveNewTag"), method = Array(RequestMethod.POST))
  def moveNewTag(@RequestParam tagId: Long, @RequestParam parentTag: Long) = {
    logger.debug(s"move new tag : $tagId, $parentTag")
    tagService.moveToTag(tagId, Option(parentTag))
    "redirect:/admin/tags"
  }

  def this() = this(null)
}